### ISSUE TITLE
Catch exceptions from formatters

### ISSUE THREAD

**Issue by moozzyk** [aspnet/SignalR#62]
Since formatters are pluggable they can fail in unexpected ways bringing the server down e.g:

```
Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.AspNetCore.SignalR.JsonNetInvocationAdapter.<>c__DisplayClass2_0.<ReadInvocationDescriptorAsync>b__0() i
n C:\source\SignalR-Core\src\Microsoft.AspNetCore.SignalR\JsonNetInvocationAdapter.cs:line 40
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.AspNetCore.SignalR.HubEndPoint`2.<DispatchMessagesAsync>d__9.MoveNext() in C:\source\SignalR-Core\src\Mi
crosoft.AspNetCore.SignalR\HubEndPoint.cs:line 105
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.AspNetCore.SignalR.HubEndPoint`2.<OnConnectedAsync>d__8.MoveNext() in C:\source\SignalR-Core\src\Microso
ft.AspNetCore.SignalR\HubEndPoint.cs:line 77
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNetCore.SignalR.HubEndPoint`2.<OnConnectedAsync>d__8.MoveNext() in C:\source\SignalR-Core\src\Microso
ft.AspNetCore.SignalR\HubEndPoint.cs:line 95
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.AspNetCore.Sockets.HttpConnectionDispatcher.<>c__DisplayClass4_0`1.<<ExecuteAsync>b__0>d.MoveNext() in C
:\source\SignalR-Core\src\Microsoft.AspNetCore.Sockets\HttpConnectionDispatcher.cs:line 116
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

We need to catch exceptions and handle them.

**Comment by moozzyk** [comment]
One _interesting_ thing is that for self describing protocols like JSon this condition is recoverable on the server - the payload can be ignored but for non-self describing protocols (e.g. protobuf) it is fatal since we don't know how much data (parameters) to read from the wire.

**Comment by BrennanConroy** [comment]
I was looking at the json one specifically for the null ref at https://github.com/aspnet/SignalR/commit/0a2bc19649678f0c73341cd1b9e14e7cbdd71ee1 Instead of throwing the client just gets an error stating the hub method doesn't exist.

**Comment by moozzyk** [comment]
Fixed via 8022afd3a2644237d597b6664f553871c69a1666
